run/django/test/retry.sh

```diff
@@ -0,0 +1,72 @@
+#!/usr/bin/env bash
+# Copyright 2021 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+##
+# retry.sh
+# Provides utility function commonly needed across Cloud Build pipelines to
+# retry commands on failure.
+#
+# Usage:
+# 1. Retry single command:
+#
+# ./retry.sh "CMD"
+#
+# 2. Retry with check:
+#
+# ./retry.sh "gcloud RESOURCE EXISTS?" "gcloud ACTION"
+#
+#
+# Note:
+#   $# - the number of command-line arguments passed
+#   $? - the exit value of the last command
+##
+
+# Usage: try "cmd1" "cmd2"
+# If first cmd executes successfully then execute second cmd
+runIfSuccessful() {
+  echo "running: $1"
+  $($1 > /dev/null)
+  if [ $? -eq 0 ]; then
+    echo "running: $2"
+    $($2 > /dev/null)
+  else return 1
+  fi
+}
+
+# Define max retries
+max_attempts=3;
+attempt_num=1;
+
+arg1="$1"
+arg2="$2"
+
+if [ $# -eq 1 ]
+then
+  cmd="$arg1"
+else
+  cmd="runIfSuccessful \"$arg1\" \"$arg2\""
+fi
+
+until eval $cmd
+do
+    if ((attempt_num==max_attempts))
+    then
+        echo "Attempt $attempt_num / $max_attempts failed! No more retries left!"
+        exit 1
```

Comment:
@averikitsch Update from the nodejs-docs-sample version (opinionated): should the retry script fail entirely, it should return an error state, meaning the cloud build step should fail, thus the job fails, and the test should fail. 
This should allow for setup failures to fail early, rather than steps failing later due to the earlier failure

Reply:
I'm not sure I follow. The `exit 1` should cause the script and then the build step to fail.

Reply:
Even with this change, the step will only fail if the last exit status is 1, so in theory if you use multiple retry.sh in one step, there is a chance it can fail and continue, depending on the order and success of commands.

Reply:
Good catch. Would this have to be handled in the cloud build to exit sooner?

Reply:
In the example I borrowed from, the cleanup was all done in one step. If and only if the last command failed in the step would the step fail.

Reply:
FYI I backed out these changes and updated my use of the script to correct the issues I was experiencing.